components/search/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+extern crate elasticlunr;
+#[macro_use]
+extern crate lazy_static;
+extern crate ammonia;
+
+extern crate errors;
+extern crate content;
+
+use std::collections::{HashMap, HashSet};
+use std::path::PathBuf;
+
+use elasticlunr::Index;
+use content::Section;
+
+
+pub const ELASTICLUNR_JS: &'static str = include_str!("elasticlunr.min.js");
+
+lazy_static! {
+    static ref AMMONIA: ammonia::Builder<'static> = {
+        let mut clean_content = HashSet::new();
+        clean_content.insert("script");
+        clean_content.insert("style");
+        let mut builder = ammonia::Builder::new();
+        builder
+            .tags(HashSet::new())
+            .tag_attributes(HashMap::new())
+            .generic_attributes(HashSet::new())
+            .link_rel(None)
+            .allowed_classes(HashMap::new())
+            .clean_content_tags(clean_content);
+        builder
+    };
+}
+
+
+/// Returns the generated JSON index with all the documents of the site added
+/// TODO: is making `in_search_index` apply to subsections of a `false` section useful?
+pub fn build_index(sections: &HashMap<PathBuf, Section>) -> String {
+    let mut index = Index::new(&["title", "body"]);
+
+    for section in sections.values() {
+        add_section_to_index(&mut index, section);
+    }
+
+    index.to_json()
```

Comment:
@mattico something weird is happening here. I only get `title` and `body` as key of the JSON but running the book example of mdbook gives me:

```json
 "fields": ["title", "body", "breadcrumbs"],
"pipeline": ["trimmer", "stopWordFilter", "stemmer"],
"ref": "id",
"version": "0.9.5",
"index": { // the actual index
```
Looking at the elasticlunr-rs code, those fields should be serialised but it's somehow not happening?

Reply:
mdbook creates the index with a breadcrumbs field: `let mut index = elasticlunr::Index::new(&["title", "body", "breadcrumbs"]);`. Does that answer your question?

Reply:
Not really, I just don't get the version, ref, pipeline, fields fields: the index generated by that line is just `{title: {...}, body: {...}}`. Way too late now though here, I'll have a better look tomorrow

Reply:
Ah, okay, I understand the question now. The `Index` struct is *just* the index. It doesn't include any of the configuration stuff. If you take a look at MDBook, [it creates that configuration by itself](https://github.com/rust-lang-nursery/mdBook/blob/77b4f6a940df900614d05f2d57373ea8b7b181cc/src/renderer/html_handlebars/search.rs#L159). I meant to add those structs to elasticlunr-rs since they're generally useful, but never got around to it. I'll take another look.

Reply:
Here it is: https://github.com/mattico/elasticlunr-rs/pull/9 Let me know what you think.

Reply:
Sorry I wasn't clear, it's not the search options that are missing.
The beautified searchindex.js of mdbook after removing the actual doc content for brevity is:

```js
window.search = {
    "searchoptions": {
        "bool": "AND",
        "expand": true,
        "limit_results": 20,
        "teaser_word_count": 30,
        "fields": {
            "title": {
                "boost": 2
            },
            "body": {
                "boost": 1
            },
            "breadcrumbs": {
                "boost": 2
            }
        }
    },
    "index": {
        "fields": ["title", "body", "breadcrumbs"],
        "pipeline": ["trimmer", "stopWordFilter", "stemmer"],
        "ref": "id",
        "version": "0.9.5",
        "index": {
            "breadcrumbs": {
            },
            "title": {
            },
            "body": {
            }
        },
        "documentStore": {
        }
    }
};
```
The `searchoptions` object is added by mdbook as you mentioned and I don't really care about that, I'll set it up in JS directly. Here's what `index.to_json` gives me in Gutenberg:

```js
window.searchIndex = {
    "body": {
    },
    "title": {
    }
};
```
In short I'm missing all of `fields`, `pipeline` etc from the first `index` object of mdbook.
The reason is that `to_json` does:

```rust 
pub fn to_json(&self) -> String {
        serde_json::to_string(&self.index).unwrap()
}
```
except elasticlunr expects all those fields (https://github.com/weixsong/elasticlunr.js/blob/master/lib/index.js#L63-L81). It looks like `to_json` should be:

```rust
pub fn to_json(&self) -> String {
        serde_json::to_string(&self).unwrap()
}
```
but maybe I'm missing something.

Reply:
Ah. That seems like a rather silly oversight. I'm sure there was I reason I did it that way... but that was several months ago.